Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm

```diff
@@ -3184,6 +3184,118 @@ - (void)_updateFixedColorExtensionViewFrames
 
 #endif // ENABLE(CONTENT_INSET_BACKGROUND_FILL)
 
+namespace WebKit {
+enum class WebViewDataType : uint32_t {
+    SessionStorage
+};
+}
+
+namespace WTF {
+template<> struct EnumTraitsForPersistence<WebKit::WebViewDataType> {
+    using values = EnumValues<
+        WebKit::WebViewDataType,
+        WebKit::WebViewDataType::SessionStorage
+    >;
+};
+}
+
+struct WKWebViewData {
+    std::optional<HashMap<WebCore::ClientOrigin, HashMap<String, String>>> sessionStorage;
+};
+
+- (void)fetchDataOfTypes:(WKWebViewDataType)dataTypes completionHandler:(void (^)(NSData *, NSError *))completionHandler
```

Comment:
Please refactor the code so that fetchDataOfTypes /  _fetchDataOfTypes, and restoreData / _restoreData will share the same implementation instead of having duplicate code.

Reply:
Fixed and added some error messages too.